sdks/python/apache_beam/transforms/environments.py

```diff
@@ -124,9 +124,9 @@ def __init__(self,
         dict(resource_hints) if resource_hints else {})
 
   def __eq__(self, other):
+    # don't compare artifacts since they have different hashes in their names.
     return (
-        self.__class__ == other.__class__ and
-        self._artifacts == other._artifacts
```

Comment:
this change is not obvious to me. I think we shouldn't include the staged job submission dependencies into the list of environment artifacts.

The transform environment describes the runtime  environment the transforms should have on the runner. Job submission dependencies currently do not impose any requirements on runtime environment, so they should not be a part of environment definition. Would it resolve the issue?

Reply:
>  Job submission dependencies currently do not impose any requirements on runtime environment...

Actually, I may be wrong here - we don't want to install the submission dependencies on the workers, but we still need to stage the file, in that sense the file is still part of the runtime environment, and it might be that list of staged artifacts is communicated via environment definition.

Reply:
https://github.com/apache/beam/blob/a4ee8548424b335cb3ffdf6aeda899fd1e3ba8ad/model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/beam_runner_api.proto#L1554

Reply:
also, it is reasonable to assume that if two environments have different set of dependencies, they are different. Why do we

Reply:
The `Stager.create_job_resources` is called from `environments.python_sdk_dependencies` which gets call when instantiating the environments. Other than this it is called in `sdk_container_builder`. I tried a run with `log_submission_env_dependencies=False` from environments and it didn't stage the file. Apparently, it has to be a part of it.

Reply:
Yeah, dependencies are a part of the environment definition so I don't think this line should be reverted. 

https://github.com/apache/beam/blob/a9516ba18a7569524fd4f8092caeb12ff4c02bf6/model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/beam_runner_api.proto#L1554

Reply:
Do you have a stack trace for the failure ?

The way it works now is that we stage all artifacts at job submission and each Environment carries the correct set of dependencies for that particular environment.

https://github.com/apache/beam/blob/a9516ba18a7569524fd4f8092caeb12ff4c02bf6/model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/beam_runner_api.proto#L1554

At runtime, the runner (UW for Dataflow) would look up the correct set of dependencies from the Runner API proto an serve them to the corresponding SDKs.

Reply:
So if you add a new dependencies, you probably have to update the runner API proto (before being staged) to include the new dependency in the corresponding environment(s).

Reply:
I agree that if two environment have different dependencies then they are not equal. But unit tests where we compare the two environments will fail with the addition of a new staging file. Even though this file is same (name and content), it has a a different hash in name:
```
first: [type_urn: "beam:artifact:type:file:v1"
type_payload: "\nd/var/folders/kk/jk392txd7x14sg93r8sjnzkh00tp20/T/tmpmxw3v4d9/submission_environment_dependencies.txt"
role_urn: "beam:artifact:role:staging_to:v1"
role_payload: "\n\'submission_environment_dependencies.txt"
]
second: [type_urn: "beam:artifact:type:file:v1"
type_payload: "\nd/var/folders/kk/jk392txd7x14sg93r8sjnzkh00tp20/T/tmpni8k_038/submission_environment_dependencies.txt"
role_urn: "beam:artifact:role:staging_to:v1"
role_payload: "\n\'submission_environment_dependencies.txt"
]
```

Reply:
that is a bizzare behavior. we should work on fixing comparison. I am not sure what is 'type_payload', but as far as file payloads are concerned, i think it's reasonably safe to compare file hashes; alternatively we can look and see why   the file path is different and if we should fix that instead. without any additional context and digging further, first options sounds more promising.

Reply:
> Do you have a stack trace for the failure ?
> 
> The way it works now is that we stage all artifacts at job submission and each Environment carries the correct set of dependencies for that particular environment.
> 
> At runtime, the runner (UW for Dataflow) would look up the correct set of dependencies from the Runner API proto an serve them to the corresponding SDKs.

Stack trace for `ExternalTransformTest` failure:
```
apache_beam/transforms/external_test.py::ExternalTransformTest::test_pipeline_generation FAILED                                                                                                [100%]

============================================================================================== FAILURES ==============================================================================================
___________________________________________________________________________ ExternalTransformTest.test_pipeline_generation ___________________________________________________________________________

self = <apache_beam.transforms.external_test.ExternalTransformTest testMethod=test_pipeline_generation>

    def test_pipeline_generation(self):
      pipeline = beam.Pipeline()
      _ = (
>         pipeline
          | beam.Create(['a', 'b'])
          | beam.ExternalTransform(
              'beam:transforms:xlang:test:prefix',
              ImplicitSchemaPayloadBuilder({'data': '0'}),
              expansion_service.ExpansionServiceServicer()))

apache_beam/transforms/external_test.py:181: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
apache_beam/pvalue.py:137: in __or__
    return self.pipeline.apply(ptransform, self)
apache_beam/pipeline.py:712: in apply
    pvalueish_result = self.runner.apply(transform, pvalueish, self._options)
apache_beam/runners/runner.py:203: in apply
    return self.apply_PTransform(transform, input, options)
apache_beam/runners/runner.py:207: in apply_PTransform
    return transform.expand(input)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ExternalTransform(PTransform) label=[ExternalTransform(beam:transforms:xlang:test:prefix)] at 0x11feb4d90>, pvalueish = <PCollection[Create/Map(decode).None] at 0x11feb4700>

    def expand(self, pvalueish):
      # type: (pvalue.PCollection) -> pvalue.PCollection
      if isinstance(pvalueish, pvalue.PBegin):
        self._inputs = {}
      elif isinstance(pvalueish, (list, tuple)):
        self._inputs = {str(ix): pvalue for ix, pvalue in enumerate(pvalueish)}
      elif isinstance(pvalueish, dict):
        self._inputs = pvalueish
      else:
        self._inputs = {'input': pvalueish}
      pipeline = (
          next(iter(self._inputs.values())).pipeline
          if self._inputs else pvalueish.pipeline)
      context = pipeline_context.PipelineContext(
          component_id_map=pipeline.component_id_map)
      transform_proto = beam_runner_api_pb2.PTransform(
          unique_name=pipeline._current_transform().full_label,
          spec=beam_runner_api_pb2.FunctionSpec(
              urn=self._urn, payload=self._payload))
      for tag, pcoll in self._inputs.items():
        transform_proto.inputs[tag] = context.pcollections.get_id(pcoll)
        # Conversion to/from proto assumes producers.
        # TODO: Possibly loosen this.
        context.transforms.put_proto(
            '%s_%s' % (self._IMPULSE_PREFIX, tag),
            beam_runner_api_pb2.PTransform(
                unique_name='%s_%s' % (self._IMPULSE_PREFIX, tag),
                spec=beam_runner_api_pb2.FunctionSpec(
                    urn=common_urns.primitives.IMPULSE.urn),
                outputs={'out': transform_proto.inputs[tag]}))
      output_coders = None
      if self._type_hints.output_types:
        if self._type_hints.output_types[0]:
          output_coders = dict(
              (str(k), context.coder_id_from_element_type(v))
              for (k, v) in enumerate(self._type_hints.output_types[0]))
        elif self._type_hints.output_types[1]:
          output_coders = {
              k: context.coder_id_from_element_type(v)
              for (k, v) in self._type_hints.output_types[1].items()
          }
      components = context.to_runner_api()
      request = beam_expansion_api_pb2.ExpansionRequest(
          components=components,
          namespace=self._external_namespace,
          transform=transform_proto,
          output_coder_requests=output_coders)
    
      expansion_service = _maybe_use_transform_service(
          self._expansion_service, pipeline.options)
    
      with ExternalTransform.service(expansion_service) as service:
        response = service.Expand(request)
        if response.error:
          raise RuntimeError(response.error)
        self._expanded_components = response.components
        if any(env.dependencies
               for env in self._expanded_components.environments.values()):
          self._expanded_components = self._resolve_artifacts(
              self._expanded_components,
>             service.artifact_service(),
              pipeline.local_tempdir)
E         AttributeError: 'ExpansionServiceServicer' object has no attribute 'artifact_service'

apache_beam/transforms/external.py:736: AttributeError
```

Reply:
> that is a bizzare behavior. we should work on fixing comparison. I am not sure what is 'type_payload', but as far as file payloads are concerned, i think it's reasonably safe to compare file hashes; alternatively we can look and see why the file path is different and if we should fix that instead. without any additional context and digging further, first option sounds more promising.

File paths are different because we are ultimately calling `python_sdk_dependencies(options)` which creates a unique temp directory with `tmp_dir = tempfile.mkdtemp()` every time